add support to use the entire file

psifertex opened the pull request:
Fixes the problems described in #2.

I'm not sure whether just switching entirely to the new method is better (maybe not for large files?) so for now I just register two menu items, but it's a pretty quick implementation. Feel free to change any way you like or let me know if you have other preferences.

paulfariello commented:
Thanks for the PR. I'm doing a quick review right now.
I agree on giving the user the choice of decompiling only a function or the full binary.

paulfariello commented on __init__.py:
```diff
@@ -30,10 +30,10 @@ def __init__(self, view, function):
-        self.cmdline.append('raw')
+        self.cmdline.append(mode)
 
+        self.cmdline = ['retdec-decompiler.sh']
```
You must init cmdline before appending `'-m'` and `mode`. Better remove your last commit.

paulfariello commented on __init__.py:
```diff
@@ -43,6 +43,23 @@ def __init__(self, view, function):
+        os.unlink('{}.c.frontend.dsm'.format(filename))
+
+        code = self.merge_symbols(code)
+        self.render_output(code)
```
We shouldn't dup the code between the two `decompile_` functions.

paulfariello commented on __init__.py:
```diff
@@ -43,6 +43,23 @@ def __init__(self, view, function):
+            code = f.read()
+
+        os.unlink('{}.c'.format(filename))
+        os.unlink('{}.c.frontend.dsm'.format(filename))
```
We must remove filename too.

paulfariello commented on __init__.py:
```diff
@@ -60,6 +77,38 @@ def decompile_function(self):
+        self.cmdline.append('{:#x}'.format(start))
+
+        self.cmdline.append(filename)
+        return filename
```
This function is duplicated it appears twice. See just below.

paulfariello commented on __init__.py:
```diff
@@ -117,11 +166,19 @@ def render_output(self, code):
+        retdec = RetDec(view, function, "bin")
+        retdec.decompile_file()
+    except Exception as e:
+        log_error('failed to decompile function\n{}'.format(e.message))
```
Should fix the log error: `failed to decompile file`.

paulfariello commented on __init__.py:
```diff
@@ -60,6 +77,38 @@ def decompile_function(self):
+            f.write(self._view.file.raw.read(0, len(self._view.file.raw)))
+
+        self.cmdline.append('--select-ranges')
+        self.cmdline.append('{:#x}-{:#x}'.format(start, start+1))
```
Don't know if this is the right way to pass the range. But I haven't tested it and it just seems odd.

psifertex replied:
Worked in testing and looks correct according to their examples: https://github.com/avast-tl/retdec/blob/master/scripts/retdec-decompiler.sh#L84

paulfariello reviewed (changes requested):
Thanks for the PR just a few minor things to fix and I'll merge it.

paulfariello commented:
Could you rebase your PR onto develop ? I've added travis-ci and pylint automation.

psifertex commented:
Sorry about that, wasn't getting these notifications. I think I resolved all the issues. Squashed and rebased on dev.

One minor clarification -- I didn't change it to actually decompile the entire binary right now, it just passes the entire binary to retdec and specifies the specific function since otherwise a lot of important context is lost that breaks disassembly even for that function (see issue #2 for example). I just call it "full" or "slower" analysis for the time being, but it might be worth documenting in the readme what the differences are. 

Travis failed, but it doesn't look like anything I'm doing as far as I can tell?

LunNova commented on __init__.py:
```diff
@@ -60,6 +78,20 @@ def decompile_function(self):
+    def read_file(self):
+        start = self._function.start
+
+        #Doesn't just operate on the original file since in a .BNDB the original file may no longer exist
```
This line and 167 are too long so it's failing pylint

paulfariello committed:
Rebase psifertex work to support entire file decompilation

Note that we only decompile current function but still use whole binary
in order to have as much information as possible.

paulfariello commented:
I rebased your PR onto develop since there was some rewrite of the plugin to support retdec config file.
Only one question left, should we decompile on range `function.start, function.start+1` or on full function?

psifertex commented:
Thanks. Sorry I was slow to get back to this. 

I used `select-ranges` because `select-function` required knowing the function name as opposed to an address, but specifying the start address of a specific function seems to result in just that function being decompiled, so believe that's the right choice but could be wrong. 

I don't think it would hurt to do start to start+1 (though I don't think it will help, either), but "full function" is a bit problematic. Mainly because there's no real definition of what that would be. If you chose the lowest address to the highest address it's possible to have interleaved basic blocks (most compilers don't do this, but there's no reason they couldn't) so another function not desired to be decompiled could also be decompiled.
